onnx/defs/traditionalml/defs.cc

```diff
@@ -915,131 +946,195 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
           updateOutputShape(ctx, 1, {N, E});
         }));
 
-static const char* TreeEnsembleRegressor_ver3_doc = R"DOC(
-    Tree Ensemble regressor.  Returns the regressed values for each input in N.<br>
-    All args with nodes_ are fields of a tuple of tree nodes, and
-    it is assumed they are the same length, and an index i will decode the
-    tuple across these inputs.  Each node id can appear only once
-    for each tree id.<br>
-    All fields prefixed with target_ are tuples of votes at the leaves.<br>
-    A leaf may have multiple votes, where each vote is weighted by
-    the associated target_weights index.<br>
-    All fields ending with <i>_as_tensor</i> can be used instead of the
-    same parameter without the suffix if the element type is double and not float.
-    All trees must have their node ids start at 0 and increment by 1.<br>
-    Mode enum is BRANCH_LEQ, BRANCH_LT, BRANCH_GTE, BRANCH_GT, BRANCH_EQ, BRANCH_NEQ, LEAF
+// Switch some ints to storing attr in tensorproto with data type uint8
+static const char* TreeEnsembleRegressor_ver5_doc = R"DOC(
+    Tree Ensemble regressor.  Returns the regressed values for each input in a batch.
+    Inputs have dimensions `[N, F]` where `N` is the input batch size and `F` is the number of input features.
+    Outputs have dimensions `[N, num_targets]` where `N` is the batch size and `num_targets` is the number of targets, which is a configurable attribute.
+
+    The encoding of this attribute is split along interior nodes and the leaves of the trees. Notably, attributes with the prefix `nodes_*` are associated with interior nodes, and attributes with the prefix `leaf_*` are associated with leaves.
+    The attributes `nodes_*` must all have the same length and encode a sequence of tuples, as defined by taking all the `nodes_*` fields at a given position.
+
+    All fields prefixed with `leaf_*` represent tree leaves, and similarly define tuples of leaves and must have identical length.
 )DOC";
 
 ONNX_ML_OPERATOR_SET_SCHEMA(
     TreeEnsembleRegressor,
-    3,
+    5,
     OpSchema()
-        .SetDoc(TreeEnsembleRegressor_ver3_doc)
-        .Input(0, "X", "Input of shape [N,F]", "T")
-        .Output(0, "Y", "N classes", "tensor(float)")
-        .TypeConstraint(
-            "T",
-            {"tensor(float)", "tensor(double)", "tensor(int64)", "tensor(int32)"},
-            "The input type must be a tensor of a numeric type.")
-        .Attr("nodes_treeids", "Tree id for each node.", AttributeProto::INTS, OPTIONAL_VALUE)
+        .SetDoc(TreeEnsembleRegressor_ver5_doc)
+        .Input(0, "X", "Input of shape [Batch Size, Number of Features]", "T")
+        .Output(0, "Y", "Output of shape [Batch Size, Number of targets]", "T")
```

Comment:
Should we ensure the threshold and weights to have the same type in shape inference? Why not adding float16 as well and let runtimes decides whether or not they want to support this type?

Reply:
Should we add support for sparse as well?

Reply:
> Should we ensure the threshold and weights to have the same type in shape inference?

This will already be happening (transitively). We check if input type equals nodes_splits type [here](https://github.com/onnx/onnx/pull/5874/files#diff-cf90e45168824f70d9af1f90cd777a4542f2637bb4f3871542679c1a0a204b82R1066) and then check that the weights type is the same as the input type [here](https://github.com/onnx/onnx/pull/5874/files#diff-cf90e45168824f70d9af1f90cd777a4542f2637bb4f3871542679c1a0a204b82R1122).

> Why not adding float16 as well and let runtimes decides whether or not they want to support this type?

I'm not opposed to this. I'm just unaware of any libraries that support this.

Reply:
> Should we add support for sparse as well?

For inputs to the decision tree?

Reply:
Let's keep it that way for the time being.